cmd/clusterctl/client/yamlprocessor/simple_processor.go

```diff
@@ -49,52 +51,111 @@ func (tp *SimpleProcessor) GetTemplateName(_, flavor string) string {
-	tmp := string(rawArtifact)
-	var err error
 	var missingVariables []string
-	for _, key := range variables {
-		val, err := variablesGetter(key)
-		if err != nil {
-			missingVariables = append(missingVariables, key)
+	// keep track of missing variables to return as error later
+	for name, hasDefault := range variables {
+		_, err := variablesClient(name)
+		// add to missingVariables list if the variable does not exist in the
+		// variablesClient AND it does not have a default value
+		if err != nil && !hasDefault {
+			missingVariables = append(missingVariables, name)
 			continue
 		}
-		exp := regexp.MustCompile(`\$\{\s*` + regexp.QuoteMeta(key) + `\s*\}`)
-		tmp = exp.ReplaceAllLiteralString(tmp, val)
 	}
+
 	if len(missingVariables) > 0 {
-		err = errors.Errorf("value for variables [%s] is not set. Please set the value using os environment variables or the clusterctl config file", strings.Join(missingVariables, ", "))
+		return rawArtifact, &errMissingVariables{missingVariables}
+	}
+
+	tmp, err = envsubst.Eval(tmp, func(in string) string {
+		v, _ := variablesClient(in)
+		return v
+	})
+	if err != nil {
+		return rawArtifact, err
 	}
 
 	return []byte(tmp), err
 }
 
-// variableRegEx defines the regexp used for searching variables inside a YAML
-var variableRegEx = regexp.MustCompile(`\${\s*([A-Z0-9_$]+)\s*}`)
+type errMissingVariables struct {
+	Missing []string
+}
 
-func inspectVariables(data []byte) []string {
-	variables := sets.NewString()
-	match := variableRegEx.FindAllStringSubmatch(string(data), -1)
+func (e *errMissingVariables) Error() string {
+	sort.Strings(e.Missing)
+	return fmt.Sprintf(
+		"value for variables [%s] is not set. Please set the value using os environment variables or the clusterctl config file",
+		strings.Join(e.Missing, ", "),
+	)
+}
 
-	for _, m := range match {
-		submatch := m[1]
-		if !variables.Has(submatch) {
-			variables.Insert(submatch)
-		}
+// legacyVariableRegEx defines the regexp used for searching variables inside a YAML.
+// It searches for variables with the format ${ VAR}, ${ VAR }, ${VAR }
+var legacyVariableRegEx = regexp.MustCompile(`(\${(\s+([A-Za-z0-9_$]+)\s+)})|(\${(\s+([A-Za-z0-9_$]+))})|(\${(([A-Za-z0-9_$]+)\s+)})`)
+var whitespaceRegEx = regexp.MustCompile(`\s`)
+
+// inspectVariables parses through the yaml string and modifies it replacing
+// variables with the format ${ VAR}, ${ VAR }, ${VAR } to ${VAR}. This is
+// done to maintain backwards compatibility.
+// It returns back the following:
+// - modified yaml string (if there were variables with spaces),
+// - map of variable names and if they have default values.
+// - error if we couldn't parse the yaml
+func inspectVariables(data string) (string, map[string]bool, error) {
+	data = legacyVariableRegEx.ReplaceAllStringFunc(data, func(o string) string {
+		return whitespaceRegEx.ReplaceAllString(o, "")
+	})
+	variables := make(map[string]bool)
+	t, err := parse.Parse(data)
+	if err != nil {
+		return data, nil, err
```

Comment:
Unfortunately, if there is bad syntax that `drone/envsubst` cannot parse (example `${foo$bar}`) it returns an unhelpful error called `bad substitution`. 
This is not helpful to the user.

I'm looking to see if I can submit a PR that would make this more useful. Perhaps a position number or partial scan of rune.